example/quadratic.f90

```diff
@@ -0,0 +1,151 @@
+program quadratic_fit
+  use nf, only: dense, input, network
+  implicit none
+  type(network) :: net
+  real, dimension(:), allocatable :: x, y
+  integer, parameter :: num_iterations = 100000
+  integer, parameter :: test_size = 30
+  real, dimension(:), allocatable :: xtest, ytest, ypred
+  integer :: i, n, batch_size
+  real :: learning_rate
+
+  print '("Fitting quadratic function")'
+  print '(60("="))'
+
+  net = network([ &
+    input(1), &
+    dense(3), &
+    dense(1) &
+  ])
+
+  call net % print_info()
+
+  allocate(xtest(test_size), ytest(test_size), ypred(test_size))
+  xtest = [(i - 1) * 2 / test_size, i = 1, test_size]
+  ytest = (xtest**2 / 2 + xtest / 2 + 1) / 2
+
+  ! x and y as 1-D arrays
+  allocate(x(num_iterations), y(num_iterations))
+
+  ! Generating the dataset
+  do i = 1, num_iterations
+    call random_number(x(i))
+    x(i) = x(i) * 2
+    y(i) = (x(i)**2 / 2 + x(i) / 2 + 1) / 2
+  end do
+
+  ! optimizer and learning rate
+  learning_rate = 0.01
+  batch_size = 10
+
+
+  ! SGD optimizer
+  call sgd_optimizer(net, x, y, learning_rate, num_iterations)
+
+  ! Batch SGD optimizer
+  call batch_sgd_optimizer(net, x, y, learning_rate, num_iterations)
+
+  ! Mini-batch SGD optimizer
+  call minibatch_sgd_optimizer(net, x, y, learning_rate, num_iterations, batch_size)
+
+  ! Calculate predictions on the test set
+  ypred = [(net % predict([xtest(i)]), i = 1, test_size)]
+
+  ! Print the mean squared error
+  print '(i0,1x,f9.6)', num_iterations, sum((ypred - ytest)**2) / size(ypred)
+
+  deallocate(x, y, xtest, ytest, ypred)
+end program quadratic_fit
```

Comment:
You may want to put the `end program quadratic_fit` statement at the very end of the program (after subroutine definitions), and place a `contains` here so that the subroutines that follow get the program scope.